src/utilities/mouse_utils.py

```diff
@@ -1,40 +1,73 @@
 import numpy as np
+import random as rd
+from typing import Callable, Union
 
 
-class MouseUtils():
+class MouseUtils:
 
-    def move_to(self, point: tuple, duration=0.3, variance=0):
+    @staticmethod
+    def move_to(point: tuple, duration: float = 0.3, duration_variance: float = 0, time_variance: float = 0,
+                tween: Union[str, Callable] = None):
         '''
         Moves mouse to a point on screen with a random movement pattern.
         Args:
             point: x, y tuple of the destination point
             duration: duration of the movement
-            variance: maximum pixel variance in final x and y position
+            duration_variance: maximum pixel variance in final x and y position
+            time_variance: the variance absolute of duration. This number is always positive (gaussian with mean 0)
+            tween: a mouse movement object from pyautogui.
+                                If None, easeInOutSine is default.
+                                If 'rand' then a selection of random movements will be used
+                                else enter a callable pyautogui function
         '''
-        # tweens = [pag.easeOutBounce, pag.easeInBounce, pag.easeInBack,
-        #           pag.easeInCirc, pag.easeInCubic, pag.easeInElastic,
-        #           pag.easeInExpo, pag.easeInOutBounce, pag.easeInOutQuad,
-        #           pag.easeInOutQuart, pag.easeInQuart, pag.easeInQuint,
-        #           pag.easeInOutBack, pag.easeOutSine]
-        # tween = tweens[np.random.randint(0, len(tweens))]
-        tween = pag.easeInOutSine
+        if tween is None:
+            tween = pag.easeInOutSine
+        elif tween == 'rand':
+            tween = rd.choice([pag.easeOutBounce, pag.easeInBounce, pag.easeInBack,
+                               pag.easeInCirc, pag.easeInCubic, pag.easeInElastic,
+                               pag.easeInExpo, pag.easeInOutBounce, pag.easeInOutQuad,
+                               pag.easeInOutQuart, pag.easeInQuart, pag.easeInQuint,
+                               pag.easeInOutBack, pag.easeOutSine])
+        else:
+            if not isinstance(tween, Callable):
+                raise TypeError('mouse_movement must be a callable function. Use None for default or "rand" for random')
         x, y = point
-        if variance != 0:
-            x += np.random.randint(-variance, variance)
-            y += np.random.randint(-variance, variance)
-        pag.moveTo(x, y, duration=duration, tween=tween)
+        if duration_variance != 0:
+            x += np.random.randint(-duration_variance, duration_variance)
+            y += np.random.randint(-duration_variance, duration_variance)
+        pag.moveTo(x, y, duration=duration + np.abs(rd.gauss(0, time_variance)), tween=tween)
 
-    def move_rel(self, x, y, duration=0.3, variance=0):
+    def move_rel(self, x: int, y: int, duration: float = 0.3, duration_variance: float = 0, time_variance: float = 0,
```

Comment:
I _think_ that since we are gonna move away from `MouseUtils` being instanced within the `Bot` class, we no longer need the `self` param - so I think we can remove it like in the above function.

Reply:
Edit: Actually it seems we do need it so we can reference the `move_to` function... Standby

Reply:
Okay, one final thing I can think of...

The `Bot` class (bot.py) uses an instance of the `MouseUtils` class as a property. I'm pretty sure this is a non-issue since from my brief testing everything seemed to work as expected -- however, since we are able to instantiate `MouseUtils`, I suppose it isn't a truly static class. This might just be my Python newbiness showing - for all I know, Python might not even enforce static class rules like some other languages. If that's the case, we should leave it as is.

![Code_NJPskyf9rD](https://user-images.githubusercontent.com/44652363/191585359-ed9b90f8-790a-4fb3-b350-c30e51b2acfc.png)

I was thinking that if we changed the line to `self.mouse = MouseUtils # excluding the parentheses`, that would make more sense, but _that_ would cause issues with the `self` param in `move_rel`, from my understanding... So maybe it's best to just leave it haha.

What do you think?

Reply:
I think it makes sense to leave it. The mouse tools only control one cursor, so the user shouldn't need to instantiate multiple class instances. But there might be a use case I'm missing!